src/block/stdio_executor.rs

```diff
@@ -202,28 +241,118 @@ impl<B: Backend> StdIoBackend<B> {
+                    // put several segments together with 1 DISCARD/WRITE ZEROES command.`, so it
+                    // seems pretty safe to drop this loop and assume one descriptor -> one struct.
+                    while available_bytes >= mem::size_of::<DiscardWriteZeroesData>() {
+                        let segment = mem
+                            .read_obj::<DiscardWriteZeroesData>(crt_addr)
+                            .map_err(Error::Read)?;
+                        self.handle_discard_write_zeroes(&segment, request.request_type())?;
+                        crt_addr = mem
+                            .checked_offset(crt_addr, mem::size_of::<DiscardWriteZeroesData>())
+                            .ok_or_else(|| {
+                                Error::GuestMemory(GuestMemoryError::InvalidGuestAddress(crt_addr))
+                            })?;
+                        available_bytes -= mem::size_of::<DiscardWriteZeroesData>();
+                    }
+                }
+            }
             RequestType::Unsupported(t) => return Err(Error::Unsupported(t)),
         };
 
         Ok(bytes_from_dev)
     }
+
+    fn handle_discard_write_zeroes(
+        &mut self,
+        segment: &DiscardWriteZeroesData,
+        request_type: RequestType,
+    ) -> Result<u32> {
+        let sector = segment.sector;
+        let num_sectors = segment.num_sectors;
+        let flags = segment.flags;
+
+        // For Discard, unmap bit (the least significant bit from segment flags) MUST be 0, for
+        // Write Zeroes it can be either 0 or 1.
+        // The other bits are reserved and MUST not be set (for both request types).
+        // If any of these conditions are not met, status must be set to VIRTIO_BLK_S_UNSUPP.
+        let valid_flags = if request_type == RequestType::WriteZeroes {
+            VIRTIO_BLK_WRITE_ZEROES_FLAG_UNMAP
+        } else {
+            0
+        };
+        if (flags & !valid_flags) != 0 {
+            return Err(Error::InvalidFlags);
+        }
+
+        let offset = sector
+            .checked_shl(u32::from(SECTOR_SHIFT))
+            .ok_or(Error::InvalidAccess)?;
+        let length = u64::from(num_sectors)
+            .checked_shl(u32::from(SECTOR_SHIFT))
+            .ok_or(Error::InvalidAccess)?;
+        self.check_access(num_sectors as u64, sector)?;
+
+        if request_type == RequestType::Discard {
+            // Since Discard is just a hint and some filesystems may not implement
+            // FALLOC_FL_PUNCH_HOLE, ignore punch_hole() errors.
+            let _ = self.inner.punch_hole(offset, length);
+        } else {
+            // If unmap is set, try at first to punch a hole, if it fails, fall back to just
+            // writing zeroes.
+            // After a write zeroes command is completed, reads of the specified ranges of sectors
+            // MUST return zeroes, independent of unmap value.
+            if flags & VIRTIO_BLK_WRITE_ZEROES_FLAG_UNMAP != 0 {
```

Comment:
Hmm, can we replace this `if/else` statement with something like:

```rust
if flags & VIRTIO_BLK_WRITE_ZEROES_FLAG_UNMAP == 0 
|| self.inner.punch_hole(offset, length).is_err() {
    self.inner
        .write_all_zeroes_at(offset, length as usize)      
        .map_err(Error::DiscardWriteZeroes)?;    
}
Ok(0)
```
Kinda wanted to avoid repeating the `write_all_zeroes` call, but pls ignore if it actually reduces clarity.